psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1791,598 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+        point_func->set_pointers(Da);
+    } else {
+        Db = wfn_->Db_subset("AO");
+        point_func = (std::shared_ptr<PointFunctions>) (std::make_shared<UKSFunctions>(basisset_, total_points, nbf));
+        point_func->set_pointers(Da, Db);
+    }
+
+    std::vector<std::shared_ptr<BlockOPoints>> blocks = grid->blocks();
+    int running_points = 0;
+
+    // Coordinates, weights, and rho (molecular electron density) at each grid point
+    std::vector<double> x_points(total_points, 0.0);
+    std::vector<double> y_points(total_points, 0.0);
+    std::vector<double> z_points(total_points, 0.0);
+    std::vector<double> weights(total_points, 0.0);
+    std::vector<double> rho(total_points, 0.0);
+
+    for (int b = 0; b < blocks.size(); b++) {
+        std::shared_ptr<BlockOPoints> block = blocks[b];
+        SharedVector rho_block;
+        std::size_t num_points = block->npoints();
+
+        point_func->set_max_functions(block->local_nbf());
+        point_func->set_max_points(num_points);
+        point_func->compute_points(block);
+        rho_block = point_func->point_values()["RHO_A"];
+
+        if (!same_dens_) {
+            rho_block->add(point_func->point_values()["RHO_B"]);
+        }
+
+        double* x = block->x();
+        double* y = block->y();
+        double* z = block->z();
+        double* w = block->w();
+
+        for (int p = 0; p < num_points; p++) {
+            x_points[running_points + p] = x[p];
+            y_points[running_points + p] = y[p];
+            z_points[running_points + p] = z[p];
+            weights[running_points + p] = w[p];
+            rho[running_points + p] = rho_block->get(p);
+        }
+
+        running_points += num_points;
+    }
+
+
+    //Calculates Number of Electrons Based on the Density Obtained using Grid Points
+    double rho_num_electrons = 0.0;
+    for (int p = 0; p < total_points; p++) {
+        rho_num_electrons += weights[p] * rho[p];
+    }
+    if (print_output) outfile->Printf("  NUMBER OF ELECTRONS (GRID DENSITY): %8.5f\n", rho_num_electrons);
+
+    //Checks to see if the number of electrons calculated using the grid densities matches actual number of electrons in molecule
+    double mol_charge = (double) (mol->molecular_charge());
+    double mol_nuc_charge = 0.0;
+
+    for (int a = 0; a < num_atoms; a++) {
+        mol_nuc_charge += mol->fZ(a);
+    }
+    double mol_electron_count = mol_nuc_charge - mol_charge;
+
+    double electron_diff = fabs(mol_electron_count - rho_num_electrons);
+
+    if (print_output) outfile->Printf("  ELECTRON DIFFERENCE: %8.5f\n\n", electron_diff);
+
+    if (electron_diff > 0.001) {
+        throw PsiException("Number of electrons calculated using the grid (" + std::to_string(round(100000.0 * rho_num_electrons) / 100000.0) + ") does not match the number of electrons in molecule. Try increasing the number of radial or spherical points (mbis_radial_points and mbis_spherical_points options).", __FILE__, __LINE__);
```

Comment:
My concern with a warning is that it would get buried in the output file. On the other hand, I can see how throwing an error might be too disruptive, particularly if the user is okay with kind of wrong charges.

Reply:
A good idea would be to let the user set the electron_difference_tolerance as an option, default will be 0.001

Reply:
We could set that option as an expert option, and give a warning/error if the tolerance is too loose

Reply:
I'm coming around to thinking of this as like the UHF spin-contamination or t1 amplitudes indicator. The integration must get printed so users can judge the quality of their calc, but it doesn't error the job. (or maybe only errors if outrageous.) It'd be good to know how missing electrons correlate with bad charges, which Andy's current grid testing will help with, too.

Reply:
I decided to only throw an exception if the electron difference is egregious (> 0.1 electrons), else (> 0.001 electrons), I throw a warning